### ISSUE TITLE
FR: Kubernetes operator: make it possible to resolve Tailnet FQDNs from within the Kubernetes cluster

### ISSUE THREAD

**Issue by irbekrm** [tailscale/tailscale#10499]
### What are you trying to do?

Today, tailnet services can be exposed to Kubernetes cluster workloads using [cluster egress proxies](https://tailscale.com/kb/1236/kubernetes-operator/?q=operator#cluster-egress) via the Tailscale Operator.
Once an egress proxy has been set up in a Kubernetes cluster, any cluster workload can access the tailnet service by referring to it via the [DNS name of the Kubernetes `Service`](https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#services) that we create for the egress proxy. 

However, in some cases, it might be required that the workload can refer to the tailnet service using its [MagicDNS name](https://tailscale.com/kb/1081/magicdns/)
There are two known use cases, a combination of both is possible:
- For a cluster workload to access a tailnet service over TLS (assuming [Tailscale HTTPS](https://tailscale.com/kb/1153/enabling-https/?q=HTTPS)) the workload needs to use the MagicDNS name (because that's what the Tailscale HTTPS cert is issued for).
- In some cases multiple components, some of which might be running in cluster and some might be user devices on tailnet, need to be configured to use the same URL to access a service, see https://github.com/tailscale/tailscale/issues/10027#issuecomment-1827763853

### How should we solve this?

A possible solution was implemented in https://github.com/tailscale/tailscale/pull/9201

We could:
- deploy a simple nameserver in the cluster
- when a Tailscale operator creates an egress proxy it could also update the nameserver with an A record for the MagicDNS name to the IP address of a Kubernetes `Service` that points to the egress proxy

Users would have to:
- update their CoreDNS/KubeDNS config with a [stub nameserver record](https://kubernetes.io/docs/tasks/administer-cluster/dns-custom-nameservers/#configuration-of-stub-domain-and-upstream-nameserver-using-coredns) for their tailnet's subdomain to the nameserver that we deployed.

Any cluster workload would then be able to access the exposed tailnet service via MagicDNS name.

### What is the impact of not solving this?

Folks who need to access access tailnet services over HTTPS would need to manually configure DNS to make this possible.
Folks who need to access workloads exposed via tailscale ingress from within the cluster, might not be able to do this at all.

**Comment by irbekrm** [comment]
Currently this is being worked on in #11017 , #11019 . #11036 and potentially implementation of #10709 would also be required to solve the case where users want to be able to access the ingress proxy backends from within the cluster over the proxy's MagicDNS name.
Here is an approximate example user flow for accessing ingress proxy from within cluster that will be possible once the above linked work is done:

**A user wants to expose a kube workload to tailnet over HTTPS and refer to it by the same hostname (the MagicDNS name of the ingress proxy) from tailnet as well as from the cluster**:
This flow requires a number of additional steps because neither our L7 nor L3 ingress proxy currently forwards non-tailscale traffic to the backend target. See #11036 for more context.

Assume that CoreDNS is used as the cluster DNS solution.

This flow assumes that #10709 has been solved via a new `ProxyClass` custom resource.

1. Deploy the operator as usual.
2. Create a new `DNSConfig` custom resource that tells the operator to create name server resources and to create DNS records for any configured ingress/egress proxies

```yaml
apiVersion: tailscale.com/v1alpha1
kind: DNSConfig
metadata:
 name ts-dns
spec:
  nameserver: {}
```

3. Wait for the `DNSConfig` to be ready and display nameserver’s Service IP:

```bash
$ kubectl get dnsconfig
NAME     NAMESERVERIP
ts-dns   10.44.14.137
``` 

4. Update [Corefile](https://coredns.io/2017/07/23/corefile-explained/) to add a forward plugin for the nameserver IP:

```yaml
ts.net {
    errors
    cache 30
    forward . 10.44.14.137
}
```

5. Create a `ProxyClass` custom resource that specifies that cluster traffic received at ingress proxy should be forwarded to the backend target in cluster:

```yaml
apiVersion: tailscale.com/v1alpha1
kind: ProxyClass
metadata:
 name: ingress-proxies
spec:
  ingressForwardClusterTrafficToTarget: true
```

6. Create the desired tailscale `Ingress`. Specify that ingress-proxies `ProxyClass` should be used

```yaml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: eng
 annotations:
   tailscale.com/proxy-class: ingress-proxies
spec:
 rules:
 <backend specifications>
  tls:
 - hosts:
   - eng
 ingressClassName:tailscale
```

7. Wait for tailscale to update `Ingress` status with the `MagicDNS` name of the ingress proxy

```bash
$ kubectl get ingress eng
NAME          CLASS       HOSTS   ADDRESS                      PORTS     AGE
eng     tailscale   *       eng.<tailnet-name>.ts.net   80, 443   81s
```


Cluster workloads as well as tailnet workloads should now be able to access the ingress backends over HTTPS (and over Wireguard in case of tailnet workloads) via eng.<tailnet-id>.ts.net.

cc @flimamacedo

**Comment by saadbahir** [comment]
> Currently this is being worked on in #11017 , #11019. Here is an approximate example user flow for accessing ingress proxy from within cluster that will be possible once the above linked work is done:

> cc @flimamacedo

Hello @irbekrm 
Thank you for these details
I recently posted another issue with a similar use case
https://github.com/tailscale/tailscale/issues/11647

I want to be able to connect 2 kubernetes cluster through tailnet using MagicDNS.

Do you know whether this would be solved by your solution?

Thank you for your help

**Comment by irbekrm** [comment]
> I want to be able to connect 2 kubernetes cluster through tailnet using MagicDNS.


Replied in https://github.com/tailscale/tailscale/issues/11647#issuecomment-2043290328
I think it makes sense for that one to be a separate issue, we probably want to document that use case separately

**Comment by stennisrl** [comment]
This is very exciting! Nice work @irbekrm 

Are there any plans to allow for a pod to resolve arbitrary FQDNs on the Tailnet without the need for a matching egress service?

My primary use case is having a pod that deploys compiled binaries to machines on the Tailnet as part of a CI pipeline. I have a considerable number of machines (to the point where creating egress services for each one would be cumbersome), so it would be great to have this functionality built into the operator. Alternatively if there's a different way of accomplishing this use case, I'm all ears

**Comment by irbekrm** [comment]
Hi @stennisrl at the moment this is aimed at operator managed mechanisms that expose tailnet services to cluster.

In generall I think allowing folks to populate the nameserver with arbitrary tailnet FQDNs would probably not be very useful since the cluster workloads would anyway not be able to reach the tailnet services on their tailnet IPs (in case of egress proxies, the records that we create are MagicDNS names -> proxy Pod IPs).

> having a pod that deploys compiled binaries to machines on the Tailnet as part of a CI pipeline

I guess currently the best solution would be to deploy Tailscale in sidecar alongside this Pod. We have thought about making it easier to inject sidecars to cluster workloads for cases were proxies are not feasible (i.e via a mutating webhook), but currently there are not that many use cases for this.
Have you tried using a sidecar for this? https://github.com/tailscale/tailscale/tree/main/docs/k8s#userspace-sidecar

**Comment by stennisrl** [comment]
That totally makes sense, thanks! 

I've been playing around with getting a sidecar to work, but DNS resolution has been problematic. My cluster uses CoreDNS, so by default I can ping tailnet IPs but can't resolve any FQDNs. I did attempt to modify the corefile to forward ts.net traffic to `100.100.100.100`, but since that IP only exists within the pod, the forward fails. 

I did also try overriding the dnsConfig for the pod so that the resolv.conf is patched to use `100.100.100.100`. This worked great for resolving tailnet FQDNs but broke resolution for anything outside of that. 

If this discussion needs to be moved elsewhere let me know, don't want to hijack this thread!
